### ISSUE TITLE
Inconsistent base class detection

### ISSUE THREAD

**Issue by Rubonnek** [godotengine/godot#40584]
**Godot version:**
3.2 branch commit 7b4b83e9dcb91b44b39e8cc87df3ddecdb29b689

**OS/device including version:**
Arch Linux

**Issue description:**
Attempting to add custom nodes using `class_name` and `tool` has inconsistent behavior -- some scripts seem to extend `Reference` as the base class when this is not the case.

Before following the steps to reproduce this issue, note that `grid_background/grid_background.gd` in the attached project extends `TextureRect`. Also `grid_background/grid_texture.gd` extends `ImageTexture`, but for some reason both are sometimes detected to extend `Reference`.

**Steps to reproduce:**
1. Open `project.godot` and check out the original `_global_script_classes`:
```
_global_script_classes=[ {
"base": "Reference",
"class": "GridBackground",
"language": "GDScript",
"path": "res://grid_background/grid_background.gd"
}, {
"base": "Reference",
"class": "GridTexture",
"language": "GDScript",
"path": "res://grid_background/grid_texture.gd"
} ]
```
2. Open the Godot Editor with `godot -e`, and when it finishes loading close it. There's a chance `GridBackground` and `GridTexture` still extend `Reference` according to `project.godot`, but if you look at their script they do not. This is the case I'm referring to -- you have reproduced the issue.
3. On the other hand, if `project.godot` do not show `Reference` as the base classes, try the following.
4. Rename `grid_background` folder to `test1` for example.
5. Open and close the Godot Editor.
6. Check `project.godot` -> `GridTexture` and `GridBackground` now "extend" `Reference` at least according to that file.

**Minimal reproduction project:**
[InconsistentBaseClassDetection.zip](https://github.com/godotengine/godot/files/4957483/InconsistentBaseClassDetection.zip)

**Comment by Xrayez** [comment]
I can consistently reproduce this by renaming the folder to anything else containing those scripts. Upon saving, `project.godot` use correct "base" class as seen in scripts. The `tool` keyword is not required to reproduce this. Godot 3.2.2.stable.

After the folder renaming, when trying to modify the script to produce some parse error I get these errors popup everytime:

```
 res://test3/grid_background.gd:6 - Parse Error: Unique global class "GridBackground" already exists at path: res://test5/grid_background.gd
 modules/gdscript/gdscript.cpp:599 - Method failed. Returning: ERR_PARSE_ERROR
 res://test3/grid_texture.gd:6 - Parse Error: Unique global class "GridTexture" already exists at path: res://test5/grid_texture.gd
 modules/gdscript/gdscript.cpp:599 - Method failed. Returning: ERR_PARSE_ERROR
```

Notice different folder names. Likely different issue, but related to this one.

**Comment by Xrayez** [comment]
Yeah, the above errors are related to #35442. I've actually tried to fix it in #40634, also in hopes to fix this one as a byproduct, but to no avail.

**Comment by Rubonnek** [comment]
Commit a97b84378c2b5aee85e66948ca763413d1532178 is causing an inconsistent base class detection in my project again.

Seems like `ResourceLoader::path_remap` is not sufficient for non-exported projects.

I'll look into this again.

**Comment by akien-mga** [comment]
I guess it's probably best to revert #40781 and #41025 then if they're not working?

**Comment by Rubonnek** [comment]
@akien-mga yeah, it was better to revert those in the meantime. Thanks!

I'm still not sure what is the root cause yet, but I'll keep digging.

**Comment by Rubonnek** [comment]
The root of the problem is Godot's cache.

More specifically the file `filesystem_cache6` stored at `$HOME/.config/godot/projects/` which is written when the attached project is opened with `godot -e` for the first time is the culprit, however it turns out that subsequently opening the attached project will make Godot store this file at `$HOME/.config/godot/projects/InconsistentBaseClassDetection-<hash_value>/filesystem_cache6` instead, which then will also cause problems. Deleting this `filesystem_cache6` file at both of these locations allows Godot to detect base classes correctly.

It's important to point out that Godot's caching mechanism is not working properly when opening projects with `godot -e` directly for the first time without importing them through `Project Manager` first. This is because when using that command line, Godot will also cache other files, aside from `filesystem_cache6`,  outside of their corresponding project folders and write them at `$HOME/.config/godot/projects` instead. Subsequently opening projects for the second time and onward, will write files to `$HOME/.config/godot/projects/<project_name>-<hash>/` instead.

I'm not familiar with the relevant code that is causing this issue.

**Comment by YuriSizov** [comment]
Global script cache behavior has been improved in 4.0 so this is likely fixed, so downgrading the milestone to 3.x.

**Comment by donn-xx** [comment]
NOTE:I am not sure if this is the right thread. 

Godot 4.1.1 stable, Ubuntu 22.04.02, Nvidia driver

I am having a hell of a bad time with these errors:
SCRIPT ERROR: Parse Error: Could not find base class <XXXX>
Could not find type "XXXX" in the current scope.

I have tried starting the project in these ways:
Godot_v4.1.1-stable_linux.x86_64 project.godot
Godot_v4.1.1-stable_linux.x86_64 -e project.godot
And starting it from the Project Manager

I have many scripts that use class_name that are @tools. It's a kind of random roll as to which class_name it will complain about each time I want to run something. 

To "fix" it, I am forced to open each script and cut/past the class_name and then ctrl+s and after that I test by ctrl+clicking the class_name. If it shows a hand, it's kind-a-sort-a okay.

I am quite stressed by this bug as I am also writing an addon and I can't imagine how to explain it to potential users who can't run the addon because they hit these errors. Quite aside from being able to just use Godot myself.

HTH

**Comment by donn-xx** [comment]
Adding some info:

If you delete the entire .godot directory and start your project, it rebuilds everything and after that the classes are recognized again.

A bit harsh, but hey..

**Comment by Eole211** [comment]
I'm encountering the same issue with the ProtonScatter plugin. It only happen in my huge game project, and inconsistently : https://github.com/HungryProton/scatter/issues/173#issuecomment-1887704094

your cut/past the class_name fix kind of works for me @donn-xx and just made my day better ! (because for my case deleting the .godot folder and reimporting all the project takes too much time, and it does not fix the bug definitely)

Next time it happens, i'll watch the .godot folder with git to see what changes during the copy/paste fix.
